stdlib/SparseArrays/src/linalg.jl

```diff
@@ -1,12 +1,54 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-import LinearAlgebra: checksquare
+import LinearAlgebra: checksquare, sym_uplo
 using Random: rand!
 
 # In matrix-vector multiplication, the correct orientation of the vector is assumed.
 const StridedOrTriangularMatrix{T} = Union{StridedMatrix{T}, LowerTriangular{T}, UnitLowerTriangular{T}, UpperTriangular{T}, UnitUpperTriangular{T}}
 const AdjOrTransStridedOrTriangularMatrix{T} = Union{StridedOrTriangularMatrix{T},Adjoint{<:Any,<:StridedOrTriangularMatrix{T}},Transpose{<:Any,<:StridedOrTriangularMatrix{T}}}
 
+function (+)(A::SparseMatrixCSC, B::Hermitian{<:Any,<:AbstractSparseMatrix})
+    C  = sparse(B)
+    C .= A .+ C
+end
+function (+)(A::SparseMatrixCSC, B::Hermitian)
+    C  = collect(B)
+    C .= A .+ C
+end
+(+)(A::Hermitian{<:Any,<:AbstractSparseMatrix}, B::SparseMatrixCSC) = B + A
+(+)(A::Hermitian, B::SparseMatrixCSC) = B + A
+
+function +(A::Hermitian{<:Any, <:SparseMatrixCSC}, B::Hermitian{<:Any, <:SparseMatrixCSC})
+    if A.uplo == B.uplo
+        Hermitian(parent(A) + parent(B), sym_uplo(A.uplo))
+    elseif A.uplo == 'U'
+        Hermitian(parent(A) + permutedims(parent(B)), :U)
+    else
+        Hermitian(permutedims(parent(A)) + parent(B), :U)
+    end
+end
+
+function (+)(A::SparseMatrixCSC, B::Symmetric{<:Any,<:AbstractSparseMatrix})
+    C  = sparse(B)
+    C .= A .+ C
+end
+function (+)(A::SparseMatrixCSC, B::Symmetric)
+    C  = collect(B)
+    C .= A .+ C
+end
+(+)(A::Symmetric{<:Any,<:AbstractSparseMatrix}, B::SparseMatrixCSC) = B + A
+(+)(A::Symmetric, B::SparseMatrixCSC) = B + A
+
+function +(A::Symmetric{<:Any, <:SparseMatrixCSC}, B::Symmetric{<:Any, <:SparseMatrixCSC})
+    if A.uplo == B.uplo
+        Symmetric(parent(A) + parent(B), sym_uplo(A.uplo))
+    elseif A.uplo == 'U'
+        Symmetric(parent(A) + permutedims(parent(B)), :U)
+    else
+        Symmetric(permutedims(parent(A)) + parent(B), :U)
+    end
+end
```

Comment:
I'm thinking that I should replace the separate `Hermitian` and `Symmetric` parts with
```julia
for T in (:Symmetric, :Hermitian)
    @eval begin
        ...
    end
end
```
because this would cut down on code duplication. Alternatively, I could define
```julia
SymOrHerm{T, Store} = Union{Symmetric{T, Store}, Hermitian{T, Store}}
```
and then define the methods on that.

Reply:
I think the for-loop pattern is common and a bit more appropriate, because you'll need both the specific constructor when returning. Just in case, you can use an already existing 
```julia
const HermOrSym{T,        S} = Union{Hermitian{T,S}, Symmetric{T,S}}
```
no need to define anything new.

Reply:
Okay, I did the `eval` for loop, but it's failing to build locally for some reason related to the `eval`, though it works if I do the `eval` in the REPL. Pushing my changes to see if someone can spot what went wrong.